From 23668cec3810b62b12c053196b83cecebc225c6d Mon Sep 17 00:00:00 2001
From: Weiwei Wang <weiwei.wang@windriver.com>
Date: Thu, 20 Jan 2011 17:30:03 +0800
Subject: [PATCH 26/28] net/dpa: fix incorrect skb headroom size

Ethernet bonding test triggers incorrect skb headroom setup of dpaa
which finally  causes ethernet port malfuntional. NET_IP_ALIGN should
be added here.

Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>
---
 drivers/net/dpa/dpaa_eth.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/net/dpa/dpaa_eth.c b/drivers/net/dpa/dpaa_eth.c
index d0223c0..eb3c934 100644
--- a/drivers/net/dpa/dpaa_eth.c
+++ b/drivers/net/dpa/dpaa_eth.c
@@ -1105,7 +1105,8 @@ static int __hot dpa_tx(struct sk_buff *skb, struct net_device *net_dev)
 	if (headroom < sizeof(skb) + NET_IP_ALIGN) {
 		struct sk_buff *skb_new;
 
-		skb_new = skb_realloc_headroom(skb, DPA_BP_HEAD);
+		skb_new = skb_realloc_headroom(skb,
+					( DPA_BP_HEAD + NET_IP_ALIGN));
 		if (!skb_new) {
 			percpu_priv->stats.tx_errors++;
 			kfree_skb(skb);
@@ -1113,7 +1114,7 @@ static int __hot dpa_tx(struct sk_buff *skb, struct net_device *net_dev)
 		}
 		kfree_skb(skb);
 		skb = skb_new;
-		headroom = DPA_BP_HEAD;
+		headroom = DPA_BP_HEAD + NET_IP_ALIGN;
 	}
 
 	skbh = (struct sk_buff **)(skb->head + NET_IP_ALIGN);
-- 
1.6.5.2

